src/rust/src/lib.rs

```diff
@@ -1,128 +1,198 @@
+    column: &Robj,
+    col_name: &str,
+    nrow: usize,
+    output_column_names: &mut Vec<String>,
+    output: &mut [f64],
+) {
+    output_column_names.extend(
+        column
+            .levels()
+            .unwrap()
+            .skip(1)
+            .map(|level| format!("{}_{}", col_name, level)),
+    );
+    // remove the first level from all the factors
+    let level_index = column.as_integer_slice().unwrap().iter().map(|x| *x - 2);
+    // let nlevels = column.levels().unwrap().len() - 1;
+    output.fill(0.);
+    for (k, level_index) in level_index.enumerate() {
+        let row_id = k;
+        // dbg!(k, level_index);
+        let col_id: Option<usize> = level_index.try_into().ok();
+        if let Some(col_id) = col_id {
+            let linear_id = row_id + col_id * nrow;
+
+            output[linear_id] = 1.0;
         }
     }
-
-    let column_names: Vec<String> = levels.iter().skip(1)
-        .map(|level| format!("{}_{}", col_name, level))
-        .collect();
-
-    (dummy_cols, column_names)
 }
 
-fn process_double_column(column: &Robj, col_name: &str, nrow: usize) -> (Array2<f64>, Vec<String>) {
-    let float_col: Array2<f64> = Array2::from_shape_vec((nrow, 1), column.as_real_vector().unwrap()).unwrap();
-    (float_col, vec![col_name.to_string()])
+fn process_double_column(
+    column: &Robj,
+    col_name: &str,
+    output_column_names: &mut Vec<String>,
+    output: &mut [f64],
+) {
+    output_column_names.push(col_name.to_string());
+    output.copy_from_slice(column.as_real_slice().unwrap())
 }
 
-fn process_string_column(column: &Robj, col_name: &str, nrow: usize) -> (Array2<f64>, Vec<String>) {
-    let str_col: Vec<&str> = column.as_str_vector().unwrap();
-    let mut level_map: HashMap<&str, usize> = HashMap::with_capacity(str_col.len());
-    let mut levels: Vec<&str> = Vec::new();
-
-    // First pass: create level mapping
-    for &s in &str_col {
-        if !level_map.contains_key(s) {
-            level_map.insert(s, levels.len());
-            levels.push(s);
-        }
-    }
-
-    let num_levels = levels.len();
-    let mut dummy_cols = Array2::<f64>::zeros((nrow, num_levels - 1));
-
-    // Second pass: fill dummy columns
-    for (i, &s) in str_col.iter().enumerate() {
-        let level_index = level_map[s];
-        if level_index > 0 {
-            dummy_cols[[i, num_levels - level_index - 1]] = 1.0;
-        }
-    }
-
-    let column_names: Vec<String> = levels.iter().skip(1).rev()
-        .map(|&level| format!("{}_{}", col_name, level))
-        .collect();
-
-    (dummy_cols, column_names)
+fn process_string_column(
+    column: &Robj,
+    col_name: &str,
+    nrow: usize,
+    output_column_names: &mut Vec<String>,
+    output: &mut [f64],
+) {
+    // R is very efficient at converting its own strings into factors, we should use that..
+    let into_factor_by_r = R!("factor({{column}})").unwrap();
+
+    let nlevels = into_factor_by_r.levels().unwrap().len() - 1;
```

Comment:
is this where the panic comes from? I think you can use `into_factor_by_r.levels()unwrap().len()`

Reply:
Close! But that's not it entirely. I need to skip one, as that's what they do to generate the model-matrix.